mantidproject/mantid
# Reduce dependency on MatrixWorkspace in DiscusMultipleScatteringCorrection

**Description of work.**

These changes remove the use of the MatrixWorkspace class from the internal calculation logic in the DiscusMultipleScatteringCorrection algorithm and replaces it with some simple data structures for 2D data based on std::vector. The MatrixWorkspace class is now only used to read input data and save the results to the ADS at the end.

The aim of this was to more loosely couple the algorithm to Mantid so that it might be usable by a wider audience eg Scipp. Although note - there still is significant Mantid coupling (eg in the use of the Mantid Geometry framework) but thought it would be a useful exercise to see how difficult it would be to loosen the coupling in terms of the workspaces at least.

**To test:**

Main things to check are that the algorithm is functionally the same and that I've not slowed it down. The unit tests on the algorithm are pretty comprehensive so hopefully they will cover the "functionally the same" part. Re performance I suggest running this Discus calculation on a release build before\after these changes and confirm it doesn't run any slower with these changes. It should take about a minute although depends on the power of your machine. Test files here [TestFiles.zip](https://github.com/mantidproject/mantid/files/10001709/TestFiles.zip):
```
# import mantid algorithms, numpy and matplotlib

from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import math

folder = r"C:\Multiple Scattering\"

sqw = Load(Filename=os.path.join(folder,'irs17901_multi_graphite002_sqw_widerq.nxs'))

sqw = Rebin(sqw,'4,0.1,21')

inputWS = Load(Filename=os.path.join(folder,'irs17901_multi_graphite002_red.nxs'))

# default cylinder axis is "up" ie y
SetSample(InputWorkspace=inputWS,
          Geometry={'Shape': 'Cylinder', 'Height': 3.0,'Radius': 1.0,'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'He'})

DiscusMultipleScatteringCorrection(InputWorkspace=inputWS, StructureFactorWorkspace=sqw,
                                   OutputWorkspace="MuscatResultsHe", NeutronPathsSingle=200,
                                   NeutronPathsMultiple=1000, NormalizeStructureFactors=False)
```

*There is no associated issue.*

*This does not require release notes* because **there is no functional change**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
